src/groth16/tests/dummy_engine.rs

```diff
@@ -21,224 +29,298 @@ impl fmt::Display for Fr {
     }
 }
 
-impl Field for Fr {
-    fn random<R: RngCore>(rng: &mut R) -> Self {
-        Fr(Wrapping(rng.next_u32()) % MODULUS_R)
+impl From<u64> for Fr {
+    fn from(n: u64) -> Fr {
+        assert!(n < MODULUS_R.0 as u64, "not in field");
+        Fr(Wrapping(n as u32))
     }
+}
 
-    fn zero() -> Self {
-        Fr(Wrapping(0))
-    }
+impl Neg for &Fr {
+    type Output = Fr;
 
-    fn one() -> Self {
-        Fr(Wrapping(1))
+    fn neg(self) -> Fr {
+        Fr(MODULUS_R - self.0)
     }
+}
+
+impl Neg for Fr {
+    type Output = Fr;
 
-    fn is_zero(&self) -> bool {
-        (self.0).0 == 0
+    fn neg(self) -> Fr {
+        -&self
     }
+}
+
+impl Add<&Fr> for &Fr {
+    type Output = Fr;
 
-    fn square(&mut self) {
-        self.0 = (self.0 * self.0) % MODULUS_R;
+    fn add(self, rhs: &Fr) -> Fr {
+        Fr((self.0 + rhs.0) % MODULUS_R)
     }
+}
 
-    fn double(&mut self) {
-        self.0 = (self.0 << 1) % MODULUS_R;
+impl Add<Fr> for &Fr {
+    type Output = Fr;
+
+    fn add(self, rhs: Fr) -> Fr {
+        self + &rhs
     }
+}
 
-    fn negate(&mut self) {
-        if !<Fr as Field>::is_zero(self) {
-            self.0 = MODULUS_R - self.0;
-        }
+impl Add<&Fr> for Fr {
+    type Output = Fr;
+
+    fn add(self, rhs: &Fr) -> Fr {
+        &self + rhs
     }
+}
+
+impl Add<Fr> for Fr {
+    type Output = Fr;
 
-    fn add_assign(&mut self, other: &Self) {
-        self.0 = (self.0 + other.0) % MODULUS_R;
+    fn add(self, rhs: Fr) -> Fr {
+        &self + &rhs
     }
+}
 
-    fn sub_assign(&mut self, other: &Self) {
-        self.0 = ((MODULUS_R + self.0) - other.0) % MODULUS_R;
+impl Sub<&Fr> for &Fr {
+    type Output = Fr;
+
+    fn sub(self, rhs: &Fr) -> Fr {
+        self + -rhs
```

Comment:
`+ -` might be needed here, but it reads funny.

Reply:
yeah I think it is needed, definitely looks funny